sdk/servicebus/service-bus/test/internal/2048-limit-test.spec.ts

```diff
@@ -0,0 +1,201 @@
+  async function sendMessages() {
+    let current = 0;
+    while (current < numberOfMessagesToSend) {
+      const batch = await sender.createMessageBatch();
+      while (
+        batch.tryAddMessage(
+          entityName.usesSessions
+            ? { body: `message-${current}`, sessionId: TestMessage.sessionId }
+            : { body: `message-${current}` }
+        ) &&
+        current++ &&
+        current < numberOfMessagesToSend
+      );
+      await sender.sendMessages(batch);
+    }
+  }
+
+  async function receiveMessages(numberOfMessagesToReceive: number) {
+    let messages: ServiceBusReceivedMessage[] = [];
+    while (messages.length < numberOfMessagesToReceive) {
+      messages = messages.concat(await receiver.receiveMessages(50, { maxWaitTimeInMs: 3000 }));
+    }
+    chai.assert.equal(
+      messages.length,
+      numberOfMessagesToReceive,
+      "Unexpected number of messages received"
+    );
+    return messages;
+  }
+
+  describe("receiveAndDelete", () => {
+    testClientTypes.forEach((clientType) => {
+      it(
+        clientType + ": would be able to receive more than 2047 messages",
+        async function(): Promise<void> {
+          await beforeEachTest(clientType, "receiveAndDelete");
+          await sendMessages();
+          await receiveMessages(numberOfMessagesToSend);
+          await verifyMessageCount(0, entityName);
+        }
+      );
+    });
+  });
+
+  describe("peekLock", () => {
+    testClientTypes.forEach((clientType) => {
+      it(
+        clientType + ": can receive a max of 2047 messages when not being settled",
+        async function(): Promise<void> {
+          await beforeEachTest(clientType);
+          await sendMessages();
+          await receiveMessages(2047);
+          await verifyMessageCount(numberOfMessagesToSend, entityName);
+          await serviceBusClient.close();
+          serviceBusClient = createServiceBusClientForTests();
+          await verifyMessageCount(numberOfMessagesToSend, entityName);
+          receiver = await serviceBusClient.test.createReceiveAndDeleteReceiver(entityName);
+          const messages = await receiveMessages(numberOfMessagesToSend);
+          console.log(messages.length);
+          const delCount = new Array(10).fill(0, 0, 10);
+          for (const message of messages) {
+            if (message.deliveryCount) {
+              delCount[message.deliveryCount]++;
+            }
+          }
+          console.log(delCount);
+          await verifyMessageCount(0, entityName);
+          // TODO:
+          // - Close the client
+          // - Receive all the messages again
+          // - Settle the messages
+          // - Delivery count should have been 1(or incremented) for 2048 of the messages
+          // - Rest 952 messages should have zero delivery count
+          // This makes sure there is no message loss
+          // Observation:
+          // - 2050 messages have deliveryCount=1 instead of the expected 2047 messages
```

Comment:
Digger deeper, this had happened because we cannot accurately stop the receiving after 2047 messages.

More background...
With the solution of checking the buffer size for every message.. (from the investigations of testing Option 4),
though I stop receiving after 2047 messages.. Since there were credits on the link, messages were on the line though we stopped receiving.
 
I was able to see this by receiving the messages again - the delivery count was 1 for 2050 messages as opposed to 2047 that I had in my hand.

If the credit initialization itself keeps the cap, the problem would be avoided entirely(as expected). 
And that's the option-7 https://gist.github.com/HarshaNalluru/0736842ac5d4024a889fdd82be015dc1#option-7

Reply:
@ramya-rao-a @chradek @richardpark-msft